include/wil/safearrays.h

```diff
@@ -0,0 +1,889 @@
+    //!     return result;
+    //! }
+    //! ~~~~
+    //! @tparam T           The underlying datatype expected in the SAFEARRAY. It should be a scalar type, BSTR, VARIANT_BOOL, LPUNKNOWN/LPDISPATCH, etc.
+    //! @tparam err_policy  Represents the error policy for the class (error codes, exceptions, or fail fast; see @ref page_errors)
+    template <typename T, typename err_policy = err_exception_policy>
+    class safearraydata_t
+    {
+    private:
+        typedef unique_any<SAFEARRAY*, decltype(&details::SafeArrayUnaccessData), details::SafeArrayUnaccessData, details::pointer_access_noaddress> safearray_unaccess_data;
+
+    public:
+        typedef typename err_policy::result result;
+        typedef safearray_unaccess_data::pointer pointer;
+        typedef T value_type;
+        typedef value_type* value_pointer;
+        typedef const value_type* const_value_pointer;
+        typedef value_type& value_reference;
+        typedef const value_type& const_value_reference;
+        typedef value_type* iterator;
+        typedef const value_type* const_iterator;
+
+    public:
+        // Exception-based construction
+        explicit safearraydata_t(pointer psa)
+        {
+            static_assert(wistd::is_same<void, result>::value, "this constructor requires exceptions; use the create method");
+            access(psa);
+        }
+        safearraydata_t() = default;
+        safearraydata_t(safearraydata_t&&) = default;
+        ~safearraydata_t() = default;
+        safearraydata_t& operator=(safearraydata_t&&) = default;
+        safearraydata_t(const safearraydata_t&) = delete;
+        safearraydata_t& operator=(const safearraydata_t&) = delete;
+
+        result access(pointer psa)
+        {
+            return err_policy::HResult([&]()
+                {
+                    WI_ASSERT(sizeof(T) == ::SafeArrayGetElemsize(psa));
+                    details::SafeArrayAccessData(psa, m_begin);
+                    m_unaccess.reset(psa);
+                    RETURN_IF_FAILED(details::SafeArrayCountElements(m_unaccess.get(), &m_size));
+                    return S_OK;
+                }());
+        }
+
+        // Ranged-for style
+        iterator begin() WI_NOEXCEPT { WI_ASSERT(m_begin != nullptr); return m_begin; }
+        iterator end() WI_NOEXCEPT { WI_ASSERT(m_begin != nullptr); return m_begin+m_size; }
+        const_iterator begin() const WI_NOEXCEPT { WI_ASSERT(m_begin != nullptr); return m_begin; }
+        const_iterator end() const WI_NOEXCEPT { WI_ASSERT(m_begin != nullptr); return m_begin+m_size; }
+
+        // Old style iteration
+        ULONG size() const WI_NOEXCEPT { return m_size; }
+        WI_NODISCARD value_reference operator[](ULONG i) { WI_ASSERT(i < m_size); return *(m_begin + i); }
+        WI_NODISCARD const_value_reference operator[](ULONG i) const { WI_ASSERT(i < m_size); return *(m_begin +i); }
+
+        // Utilities
+        bool empty() const WI_NOEXCEPT { return m_size == ULONG{}; }

```

Comment:
Just compare to zero (or `0ul` if sign comparison is giving warnings, which I'd hope it wouldn't)

Reply:
I think I originally had the size type be a template parameter, which is why it ended up like that.  I have no problem changing it.

Reply:
I think I had the size original as a size_type typedef, which is why I wrote it like that.  No issue changing it.